test/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter_integration_test.cc

```diff
@@ -0,0 +1,128 @@
+  for (uint32_t idx = 0; idx < request_count - num_forwarded; ++idx) {
+    auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+    responses_.push_back(std::move(encoder_decoder.second));
+    codec_client_->sendData(encoder_decoder.first, 0, true);
+
+    // These will remain nullptr.
+    upstream_connections_.emplace_back();
+    upstream_requests_.emplace_back();
+  }
+
+  ASSERT_EQ(upstream_connections_.size(), upstream_requests_.size());
+  ASSERT_EQ(responses_.size(), upstream_requests_.size());
+}
+
+void AdaptiveConcurrencyIntegrationTest::respondToAllRequests(
+    const int forwarded_count, const std::chrono::milliseconds latency) {
+  ASSERT_GE(responses_.size(), static_cast<size_t>(forwarded_count));
+
+  timeSystem().sleep(latency);
+
+  for (int idx = 0; idx < forwarded_count; ++idx) {
+    respondToRequest(true);
+  }
+  while (!responses_.empty()) {
+    respondToRequest(false);
+  }
+}
+
+void AdaptiveConcurrencyIntegrationTest::respondToRequest(const bool expect_forwarded) {
+  ASSERT_EQ(upstream_connections_.size(), upstream_requests_.size());
+  ASSERT_EQ(responses_.size(), upstream_requests_.size());
+
+  if (expect_forwarded) {
+    ASSERT_NE(upstream_connections_.front(), nullptr);
+    ASSERT_NE(upstream_requests_.front(), nullptr);
+    ASSERT_TRUE(upstream_requests_.front()->waitForEndStream(*dispatcher_));
+    upstream_requests_.front()->encodeHeaders(default_response_headers_, true);
+  }
+
+  responses_.front()->waitForEndStream();
+
+  if (expect_forwarded) {
+    EXPECT_TRUE(upstream_requests_.front()->complete());
+  }
+
+  EXPECT_TRUE(responses_.front()->complete());
+
+  if (expect_forwarded) {
+    verifyResponseForwarded(std::move(responses_.front()));
+    ASSERT_TRUE(upstream_connections_.front()->close());
+    ASSERT_TRUE(upstream_connections_.front()->waitForDisconnect());
+  } else {
+    verifyResponseBlocked(std::move(responses_.front()));
+  }
+
+  upstream_connections_.pop_front();
+  upstream_requests_.pop_front();
+  responses_.pop_front();
+}
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, AdaptiveConcurrencyIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()));
+
+/**
+ * Test a single request returns successfully.
+ */
+TEST_P(AdaptiveConcurrencyIntegrationTest, TestConcurrency1) {
+  customInit();
+
+  sendRequests(2, 1);
+  respondToAllRequests(1, std::chrono::milliseconds(1));
+  test_server_->waitForCounterEq(REQUEST_BLOCK_COUNTER_NAME, 1);
+}
+
+/**
+ * Test many requests, where only a single request returns 200 during the minRTT window.
+ */
+TEST_P(AdaptiveConcurrencyIntegrationTest, TestManyConcurrency1) {
+  customInit();
+
+  sendRequests(10, 1);
+  respondToAllRequests(1, std::chrono::milliseconds(1));
+  test_server_->waitForCounterEq(REQUEST_BLOCK_COUNTER_NAME, 9);
+}
+
+/**
+ * TODO: Test the ability to increase/decrease the concurrency limit with request latencies based on
```

Comment:
Can you briefly expand on what you tried to do to get this working already so we don't lose the context? cc @jmarantz